RTC: STM32: configure U5xx boards

ajarmouni-st opened the pull request:
Adds RTC clock for U5xx based boards.
STM32 RTC does not support years before 2000 ([rtc_ll_stm32.c](https://github.com/zephyrproject-rtos/zephyr/blob/2b549daabc9d2cddda3e22c6ec1f5b27ea1997ad/drivers/rtc/rtc_ll_stm32.c#L184)). Excludes the test_y2k test on STM32 boards.
Adds stm32 test cases, configs, & overlays.

ajarmouni-st committed:
boards: st: u5xx: add RTC clock

Add RTC clock for U5xx based boards, & update doc.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
tests: drivers: rtc: rtc_api: exclude test_y2k on stm32

STM32 RTC does not support years before 2000 (rtc_ll_stm32.c:185).

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
tests: drivers: rtc: rtc_api: add stm32 test cases & configs

Add stm32 configs. Not all series support RTC calibration.
None of the stm32 SoCs support RTC alarm right now.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
board: arm: nucleo_f767zi: flashing

Add the possibility to flash nucleo-f767zi board using west
STM32CubeProgrammer runner.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

erwango commented on tests/drivers/rtc/rtc_api/src/test_y2k.c:
I can't tell if there was actually an interest in this specific y2k test in 2024, but [y2k bug](https://en.wikipedia.org/wiki/Year_2000_problem) is an actual thing. Modfying it to y2k1 doesn't make much sense IMO.
I would rather propose to have a way not to run it on STM32

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/testcase.yaml:
```diff
@@ -1,13 +1,25 @@
-    depends_on: rtc
     platform_exclude:
       - qemu_x86_64
+  drivers.rtc.rtc_api.stm32:
+    filter: CONFIG_SOC_FAMILY_STM32
+    extra_args:
+      - DTC_OVERLAY_FILE="boards/stm32-rtc.overlay"
+    platform_allow:
+      - b_u585i_iot02a
+      - nucleo_f767zi
+      - nucleo_h563zi
+      - nucleo_h723zg
+      - nucleo_u575zi_q
```
Board specific modifications should be in `boards/` rather than the testcase.yaml

Using `boards/*.overlay` and `boards/*.conf`, modifications to Kconfig and dts is possible.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/src/test_y2k.c:
```diff
@@ -10,6 +10,10 @@
 
 #include <time.h>
 
+#if defined(CONFIG_SOC_FAMILY_STM32)
```
A more scalable solution here is to add a Kconfig to exclude the Y2K test, and enable this Kconfig in the boards .conf files.

ajarmouni-st replied:
I did exclude the y2k test file in CMakeLisks, the `#ifdef` here is just an extra layer to prevent building this test for STM32.

ajarmouni-st committed:
samples: drivers: counter: alarm: add stm32-rtc test case

A specific test case for STM32 RTC is needed because of the way Counter
node is selected in preprocessing.
This is the only test case that tests STM32 RTC_Alarm currently since
its not yet supported in rtc_ll_stm32 driver, & only implemented in
counter_ll_stm32_rtc driver.
This also clarifies how STM32 RTC can be used in this test for the user.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/testcase.yaml:
```diff
@@ -10,4 +11,9 @@ tests:
     filter: dt_alias_exists("rtc")
     depends_on: rtc
     platform_exclude:
+      - b_u585i_iot02a
+      - nucleo_f767zi
+      - nucleo_h563zi
+      - nucleo_h723zg
+      - nucleo_u575zi_q
```
Excluding the added boards is not required. They can not be run, so they will only be built, which is a good test in itself :) The reason `qemu_x86_64` is excluded is because it can be run, but runs at real-time, timing out the test suite :)

ajarmouni-st replied:
Of crs, sorry, I just didn't pay attention it's exclude & not include.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/src/test_y2k.c:
```diff
@@ -10,6 +10,10 @@
 
 #include <time.h>
 
+#ifdef CONFIG_SOC_FAMILY_STM32
+#error STM32 RTC does not support years before 2000
+#endif  /* CONFIG_SOC_FAMILY_STM32 */
+
```
This really is not required, the file is already excluded in the CMakeLists.txt, with a comment stating why

ajarmouni-st committed:
boards: st: nucleo_f722ze: activate LSI clock for RTC

activate LSI clock for RTC.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

bjarki-trackunit commented:
Changes regarding RTC are good :) are the failing tests caused by this PR? if so, they need to be fixed :)

ajarmouni-st commented:
> Are the failing tests caused by this PR?

No. I pushed a fix for the failing test here https://github.com/zephyrproject-rtos/zephyr/pull/71329

ajarmouni-st commented on samples/drivers/counter/alarm/sample.yaml:
```diff
@@ -28,12 +38,11 @@ tests:
-        - "!!! Alarm !!!"
-        - "Now: [2|3]"
-    depends_on: counter
+  sample.drivers.counter.alarm.stm32_rtc:
+    # This test case is needed because when Timer Counter is available,
+    # RTC Counter will never be selected in the test.
+    # CONFIG_COUNTER_RTC_STM32 flag is enabled by default when RTC is enabled.
+    filter: CONFIG_SOC_FAMILY_STM32 and CONFIG_COUNTER_RTC_STM32
+    extra_configs:
+      # Timer Counter flag needs to be disabled for RTC node to be selected
+      - CONFIG_COUNTER_TIMER_STM32=n
```
As I mentioned in the commit message, "this is the only test case that tests STM32 RTC_Alarm currently since its not yet supported in rtc_ll_stm32 driver, & only implemented in counter_ll_stm32_rtc driver." so in my opinion it should be kept in this PR.

ajarmouni-st replied:
> We need to find a better way to avoid configuration issues

Of course, will push another PR for this.

FRASTM commented on boards/st/nucleo_f722ze/nucleo_f722ze.dts:
```diff
@@ -69,6 +69,10 @@
 	status = "disabled";
 };
 
+&clk_lsi {
```
if this PR is for stm32U5, why changing stm32F7 nucleo boards ?

FRASTM commented on tests/drivers/rtc/rtc_api/boards/b_u585i_iot02a.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_RTC_CALIBRATION=y
```
Or we could have a testcase item with EXTRA_CONFIGS for that

ajarmouni-st replied:
Not all boards support RTC calibration. Also we would need to maintain the list to insure that the calibration test is not run in multiple test cases.
